keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/LogProbGradientCalculatorTest.java

```diff
@@ -0,0 +1,128 @@
+
+public class LogProbGradientCalculatorTest {
+
+    @Test
+    public void canFindGradientOfSingleVariateGaussianWrtSelf() {
+        GaussianVertex A = new GaussianVertex(0, 1);
+        A.setValue(0.5);
+
+        LogProbGradientCalculator calculator = new LogProbGradientCalculator(ImmutableList.of(A), ImmutableList.of(A));
+
+        Map<VertexId, DoubleTensor> gradient = calculator.getJointLogProbGradientWrtLatents();
+        DoubleTensor dALogProbWrtAValue = gradient.get(A.getId());
+
+        ContinuousDistribution distribution = Gaussian.withParameters(DoubleTensor.ZERO_SCALAR, DoubleTensor.ONE_SCALAR);
+        DoubleTensor expected = distribution.dLogProb(DoubleTensor.scalar(0.5)).get(Diffs.X).getValue();
+
+        assertThat(dALogProbWrtAValue, equalTo(expected));
+    }
+
+    @Test
+    public void canFindGradientOfSingleVariantGaussianWrtSingleVariateLatent() {
+
+        GaussianVertex A = new GaussianVertex(0, 1);
+        GaussianVertex B = new GaussianVertex(A, 1);
+        B.setValue(0.5);
+
+        LogProbGradientCalculator calculator = new LogProbGradientCalculator(ImmutableList.of(B), ImmutableList.of(A));
+
+        Map<VertexId, DoubleTensor> gradient = calculator.getJointLogProbGradientWrtLatents();
+        DoubleTensor dBLogProbWrtAValue = gradient.get(A.getId());
+
+        DoubleTensor expectedDLogProbWrtA = B.dLogProb(DoubleTensor.scalar(0.5), A).get(A);
+
+        assertThat(dBLogProbWrtAValue, equalTo(expectedDLogProbWrtA));
+    }
+
+    @Test
+    public void canFindGradientOfSingleVariantGaussianWrtMultivariateLatent() {
+
+        GaussianVertex A = new GaussianVertex(new int[]{3, 2}, 0, 1);
+        GaussianVertex B = new GaussianVertex(A, 1);
+        DoubleTensor bValue = DoubleTensor.create(new double[]{0.1, 0.2, 0.3, -0.2, -0.5, 0.9}, 3, 2);
+        B.setValue(bValue);
+
+        LogProbGradientCalculator calculator = new LogProbGradientCalculator(ImmutableList.of(B), ImmutableList.of(A));
+
+        Map<VertexId, DoubleTensor> gradient = calculator.getJointLogProbGradientWrtLatents();
+        DoubleTensor dBLogProbWrtAValue = gradient.get(A.getId());
+
+        DoubleTensor expectedDLogProbWrtA = B.dLogProb(bValue, A).get(A);
+
+        assertThat(dBLogProbWrtAValue, equalTo(expectedDLogProbWrtA));
+    }
+
+    @Test
+    public void canFindGradientOfMultivariantGaussianWrtSingleVariateLatent() {
+
+        GaussianVertex A = new GaussianVertex(0, 1);
+        GaussianVertex B = new GaussianVertex(new int[]{3, 2}, A, 1);
+        DoubleTensor bValue = DoubleTensor.create(new double[]{0.1, 0.2, 0.3, -0.2, -0.5, 0.9}, 3, 2);
+        B.setValue(bValue);
+
+        LogProbGradientCalculator calculator = new LogProbGradientCalculator(ImmutableList.of(B), ImmutableList.of(A));
+
+        Map<VertexId, DoubleTensor> gradient = calculator.getJointLogProbGradientWrtLatents();
+        DoubleTensor dBLogProbWrtAValue = gradient.get(A.getId());
```

Comment:
As mentioned above, if this method returned a map of `Vertex` to `DoubleTensor` we wouldn't have to do this `getId()`